File: DSA_Template_CPP/Algo_Levenshtein.cpp
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

int levenshteinDistance(string str1, string str2) {
    int m = str1.size(); // Length of str1
    int n = str2.size(); // Length of str2
    vector<vector<int>> dp(m + 1, vector<int>(n + 1)); // DP matrix

    // Initialize first column and first row of DP table
    for (int i = 0; i <= m; i++) dp[i][0] = i;
    for (int j = 0; j <= n; j++) dp[0][j] = j;

    // Compute the distances
    for (int i = 1; i <= m; i++) {
        for (int j = 1; j <= n; j++) {
            if (str1[i - 1] == str2[j - 1]) {
                // Characters match, cost is 0
                dp[i][j] = dp[i - 1][j - 1];
            } else {
                // Characters do not match, calculate cost
                dp[i][j] = 1 + min({dp[i - 1][j],      // Deletion
                                    dp[i][j - 1],      // Insertion
                                    dp[i - 1][j - 1]});// Substitution
            }
        }
    }
    return dp[m][n]; // The Levenshtein distance is in the bottom-right cell
}

File: DSA_Template_Python/Algo_Levenshtein.py
def levenshteinDistance(str1, str2):
    m, n = len(str1), len(str2)  # Lengths of the two strings
    dp = [[0] * (n + 1) for _ in range(m + 1)]  # Initialize DP table

    # Set up the first column and row
    for i in range(m + 1):
        dp[i][0] = i
    for j in range(n + 1):
        dp[0][j] = j

    # Iterate over each character in the strings
    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if str1[i - 1] == str2[j - 1]:
                # Characters match, no operation needed
                dp[i][j] = dp[i - 1][j - 1]
            else:
                # Characters do not match, consider all operations
                dp[i][j] = 1 + min(dp[i - 1][j],    # Deletion
                                   dp[i][j - 1],    # Insertion
                                   dp[i - 1][j - 1])# Substitution

    return dp[m][n]  # The Levenshtein distance is in the bottom-right cell
